sycl/doc/extensions/DiscardQueueEvents/SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS.asciidoc

```diff
@@ -0,0 +1,187 @@
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+== Overview
+
+This extension adds `ext::oneapi::property::queue::discard_events` property for `sycl::queue`,
+by using this property the application informs a SYCL implementation that the returned event can be discarded.
+
+In cases where it will be possible, the implementation makes a decision about internal optimization -- avoiding
+the creation of low-level event from the backend. The property affects only kernel cases,
+i.e. avoid creating low-level event for submission of kernels.
+Performance improvement is the main motivation for using this property.
+
+Below is a usage example:
+[source,c++]
+----
+  sycl::property_list props{ext::oneapi::property::queue::discard_events{}
+                            property::queue::in_order{}};
+  sycl::queue Queue( props );
+
+  // some USM preparations ..
+
+  sycl::event e1, e2, e3;
+
+  // returning "invalid" events from each submission function:
+  e1 = Queue.parallel_for(NDRange, [=](nd_item<1> item){ do_smth1(); });
+
+  e2 = Queue.single_task([=](){ do_smth2(); });
+
+  e3 = Queue.submit([&](handler &CGH) { CGH.parallel_for(NDRange, [=](nd_item<1> item){ do_smth3(); }); });
+
+  Queue.wait();
+----
+
+In the example above, the application doesn't use sycl events: `e1`, `e2`, `e3`
+and is waiting for the end of work by `queue::wait()`. The returned events will be
+_invalid_ events. _invalid_ event is a `sycl::event` with a status `invalid` and the event returns from each kernel
+submission when queue is created with the `discard_events` property.
+See the description of behavior for this event below for details.
+
+Here, only those methods for the _invalid_ event are described that have behavior different from the default event behavior:
+[source,c++]
+----
+// must throw an exception with the errc::invalid error code.
+void wait();
+
+// if invalid event is passed into the function, must throw an exception with the errc::invalid error code.
+static void wait(const std::vector<event> &eventList);
+
+// must throw an exception with the errc::invalid error code.
+void wait_and_throw();
+
+// if invalid event is passed into the function, must throw an exception with the errc::invalid error code.
+static void wait_and_throw(const std::vector<event> &eventList);
+
+// must return info::event_command_status::ext_oneapi_unknown
+get_info<info::event::command_execution_status>() const;
+----

```

Comment:
Should we say anything special for `get_wait_list()`? I believe it can follow the original semantic.

Reply:
Are you suggesting that `get_wait_list()` should return normal (i.e. not "invalid") events?

Thinking about this more, it would probably be easier to just say that `get_wait_list()` throws `errc::invalid`.

Reply:
BTW, the default event returns a zero-size vector for `get_wait_list()`, So I believe the invalid event also can save the semantic, Should we change it to throwing `errc::invalid` just in case?

Reply:
Returning an empty set of events for the default-constructed event make sense because that event is always "complete".  Thus, there are never any other events to wait for.

I do not think it makes sense to return an empty set of events for an "invalid" event.  If we did, code like this would silently do the wrong thing:

```
cgh.depends_on(myevt.get_wait_list());
```

Whatever we do, code like I show above should cause an exception.  This would happen if `get_wait_list()` returns a vector of "invalid" events, or it would happen if `get_wait_list()` itself throws an exception.  Throwing an exception from `get_wait_list()` is easier, so that is probably what makes sense.

Reply:
```
cgh.depends_on(myevt.get_wait_list());
```
Interesting example, I was think about such a case:

```
sycl::queue QReg{...}; // regular queue
sycl::queue QDiscardEvents{...}; // a queue with ext::oneapi::property::queue::discard_events{}
auto RegEvent = QReg.submit(...);
auto DiscardedEvent = QDiscardEvents.submit(... CGH.depends_on(RegEvent); ...); // A commands depends on a regular event from another queue

auto WaitList = DiscardedEvent.get_wait_list(); // get_wait_list line
```

I would not expect that `get_wait_list line` returns zero event list(in general case) or list of invalid events.
So, throwing an exception from `get_wait_list` sounds reasonable.